### ISSUE TITLE
MARC records listed as source records not being used (or used fully?)

### ISSUE THREAD

**Issue by tfmorris** [internetarchive/openlibrary#9831]
### Problem

When investigating editions records with no publishers for #2119, I noticed cases where the `source_records` lists MARC records which contain publishers in the MARC 260, but it's not being added to the record

This edition: https://openlibrary.org/books/OL10298M?m=history
was imported from a rather threadbare Scriblio MARC record: https://openlibrary.org/show-records/marc_records_scriblio_net/part29.dat:6096809:617
but a later import claims to have used a much richer Columbia MARC
https://openlibrary.org/books/OL10298M/Tiricirapuram_maka%CC%84vittuva%CC%84n%CC%B2_..._Amparp_pura%CC%84n%CC%A3am?b=5&a=4&_compare=Comparer&m=diff
yet didn't pull in the publisher from there.

Additionally, the original record elided words from the title, so a search on the full title returns zero hits, but I'm not sure there's a good way to detect and correct for that case.

The second example: https://openlibrary.org/books/OL12026877M
also can't be found by title, but because it was imported from a threadbare (and incorrect) Amazon record with a typo in it.
Despite "importing" from four higher quality MARC records, all containing the correct title and a fully populated MARC 260 Publisher field, neither the missing publisher field nor the incorrect title were updated.

Before trying to guess publishers based on ISBN, the high quality metadata that's already available should be fully exploited.

### Reproducing the bug

1. Go to ...
2. Do ...

* Expected behavior:
* Actual behavior:

### Context

- Browser (Chrome, Safari, Firefox, etc):
- OS (Windows, Mac, etc):
- Logged in (Y/N):
- Environment (prod, dev, local): prod

### Breakdown

See this comment: https://github.com/internetarchive/openlibrary/issues/9831#issuecomment-2351781827

#### Requirements Checklist
Taken from https://github.com/internetarchive/openlibrary/issues/9831#issuecomment-2351781827:
* [ ] identify no-`publishers` records with MARC sources
* [ ] re-import those MARC records (once #9808 has been deployed to production)
* [ ] the records should be correctly matched, and any existing publisher metadata will be updated using the latest import code

#### Related files

*

#### Stakeholders

* @hornc 
<hr>

#### Instructions for Contributors

- Please [run these commands](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#working-on-your-branch) to ensure your repository is up to date **before** [creating a new branch](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#making-changes-and-creating-a-pull-request) to work on this issue and **each time after** pushing code to Github, because the pre-commit bot may add commits to your PRs upstream.

**Comment by tfmorris** [comment]
Should have included @hornc in the stakeholders - updated.

**Comment by seabelis** [comment]
It seems like there's a high error rate with matching these MARC imports to existing records. I'd not be in favour of modifying records based on them (but don't we already do that now?).

**Comment by scottbarnes** [comment]
#9808 may be responsive to the matching issue, but perhaps not. I am unsure the full extent of it.

**Comment by hornc** [comment]
I agree that #9808 should make existing record matching considerably better -- that fixed a longstanding issue whereby records were frequently matched just on title only (ignoring subtitle and any other metadata). These matches were made before even attempting the more sophisticated threshold matching code that exists and has tests in the codebase.

`publishers` from new records should be added to matched existing records if they are blank; (I had to search for the code I though/hoped existed):

https://github.com/internetarchive/openlibrary/blob/f64cab54045351216cd22b961691d7946ecc0a14/openlibrary/catalog/add_book/__init__.py#L834-L840


It looks like `publishers` was added to this list in Feb 2023 in this commit
https://github.com/internetarchive/openlibrary/commit/f6268b647eb8e783e0ef3f1203153a65e64c9c96

, which is after the reported example where publisher wasn't added in Aug 2022:
https://openlibrary.org/books/OL10298M/Tiricirapuram_maka%CC%84vittuva%CC%84n%CC%B2_S%CC%81ri%CC%84_Mi%CC%84n%CC%B2a%CC%84t%CC%A3cicuntaram_Pil%CC%A3l%CC%A3ai_avarkal%CC%A3_iyar%CC%B2r%CC%B2iya_Tiru_Amparp_pur?b=5&a=4&_compare=Comparer&m=diff

I believe the code does the correct thing now, but only since 2023, so there will be many examples where it has been missed. 

If we wanted to populate missing `publishers` :
* identify no-`publishers` records with MARC sources
* re-import those MARC records (once #9808 has been deployed to production)
* the records should be correctly matched, and any existing publisher metadata will be updated using the latest import code

**Comment by LeadSongDog** [comment]
Another example. This talis import led to the wrong field (245 vice 240) being used to source the edition title:

<img width="2048" height="1536" alt="Image" src="https://github.com/user-attachments/assets/b99a20cf-316d-401c-b41e-d767eb8ff239" />

<img width="2048" height="1536" alt="Image" src="https://github.com/user-attachments/assets/a9f07958-2fdb-4767-abee-bc970b192601" />

Note: on the edition record the title has since been manually corrected.

**Comment by tfmorris** [comment]
By not including any URLs you make this too hard to investigate, but, on the surface, looking at the images, the importer did the correct thing. The MARC 245 is the title and was correctly imported as such. The Talis records aren't the highest quality,  so perhaps it was miscatalogued originally, but I don't think the importer did anything wrong.

**Comment by LeadSongDog** [comment]
@tfmorris Argh, I should have caught that.  The talis did indeed transpose 240 with 245.

**Comment by mekarpeles** [comment]
**Status overview**

The checklist in the issue body (from @hornc's comment) lists three steps gated on issue #9808 being deployed to production. #9808 is now closed, and the related PR #9839 merged in September 2024. The question is whether those merged changes are sufficient to close the loop, or whether the data remediation step (re-importing no-publisher MARC records) is still outstanding.

**What #9839 fixed**

PR #9839 prevented MARC records without ISBNs from incorrectly matching ISBN-sourced "promise item" records — a significant match-quality improvement. The `publishers` field has been on the overwrite list since Feb 2023 (`add_book/__init__.py` line ~834), so re-importing a rich MARC record over a threadbare one *should* now pull in the publisher.

**What may still be open**

The original examples (editions that were "imported" from rich MARC records but still have empty `publishers`) may not have been re-imported since the fix. The data is stale, not the code. Verifying one of the original examples would clarify this:

- https://openlibrary.org/books/OL10298M — check if `publishers` is still blank
- If blank and has a MARC source_record, a targeted re-import via `openlibrary-client` should populate it now

**Suggested next step**

1. Check one of the original reported editions — if `publishers` is populated, the fix deployed correctly and this issue can be closed
2. If still blank, a scoped re-import pass targeting editions with `publishers: []` + `source_records` containing a `marc:` prefix is the remaining work

— Richy, via PAM (Open Library's Project AI Manager)

**Comment by hornc** [comment]
The record https://openlibrary.org/books/OL10298M still had blank publishers.
There are three MARC sources.

* Re-importing the LOC one, https://openlibrary.org/show-records/marc_loc_2016/BooksAll.2016.part43.utf8:13831012:624 Matched the record and made no changes.
* Re-importing the Columbia one: https://openlibrary.org/show-records/marc_columbia/Columbia-extract-20221130-011.mrc:45401045:1513 Matched and ***modified*** the record, adding the publisher as shown in this [edit](https://openlibrary.org/books/OL10298M/Tiricirapuram_maka%CC%84vittuva%CC%84n%CC%B2_S%CC%81ri%CC%84_Mi%CC%84n%CC%B2a%CC%84t%CC%A3cicuntaram_Pil%CC%A3l%CC%A3ai_avarkal%CC%A3_iyar%CC%B2r%CC%B2iya_Tiru_Amparp_pur?_compare=Compare&b=8&a=7&m=diff)

**Comment by hornc** [comment]
The other example is fixed via:  https://openlibrary.org/books/OL12026877M/For_the_Love_of_One_Anther?_compare=Compare&b=9&a=8&m=diff

This was hard to test due to persistent `460` errors that were occurring on reimport attempts.
